### DISCUSSION TITLE
Input file parsing

### DISCUSSION THREAD

**Discussion by wpbonelli** [modflowpy/pyphoenix-project#55] [Ideas]
We have discussed/prototyped using [`lark`](https://github.com/lark-parser/lark) to decode MF6 input files into Python containers and primitives (before passing these to `cattrs` for structuring).

The prototype is currently (mostly) functional. But we still need to determine how performance compares to manual decoding. This will involve benchmarking/profiling the lark prototype against flopy3. If it becomes evident that flopy3 can be optimized in fairly straightforward ways, maybe it will be useful to add a minimal hand-written decoder prototype here for better comparison.

This is related to the performance testing mentioned in https://github.com/modflowpy/pyphoenix-project/discussions/49#discussioncomment-11438175, but the focus areas are slightly different:

- here we are only concerned with load performance, and can ignore write performance
- in the other issue, we are most concerned with IO for data structures (arrays/tables/lists) specifically, where here we consider the full contents of input files, including internal and/or external data

**Comment by deltamarnix** [comment]
Sometimes I could prefer readability over performance.

- [ ] Are we concerned that load times will increase significantly?
- [ ] Are we only speaking of MF6 input files, or also MF6 output files?

**Reply by wpbonelli** [reply]
> Are we concerned that load times will increase significantly?

This will need benchmarking to make sure they don't, and rather decrease.

> Are we only speaking of MF6 input files, or also MF6 output files?

Lark is more suitable for input files than output files IMO. Input files share the same format, so we only have to write 1 grammar. Output files have a variety of formats. Maybe at some point it's worth considering for output files too, but maybe not yet.

**Comment by wpbonelli** [comment]
The main benefit provided by lark is maintainability. Instead of hand-writing a parser, a parser is generated from a grammar.

We can move forward with lark, and if performance turns out to be unacceptable, bite the bullet and hand-write something. To maximize performance the grammar should be kept as simple as possible (e.g., injecting component-specific keywords into the grammar rather than defining a grammar that is generic over all components).

Perhaps consider a prospective back-compat policy for flopy4 (e.g. dfns will be supported X years or Y releases). We can be more lenient with retrospective back-compat &mdash; upon competing an initial prototype, we can test against historic dfn versions and see how far back we can support.

In the long run, pursue standardizing input files with

- netcdf
- toml/etc

**Comment by wpbonelli** [comment]
The reason the proof of concept is so slow is the grammar is ambiguous. So lark has to use the Earley parser. For LALR the grammar can't have ambiguities. If we can get them out it should run in linear time.

**Reply by wpbonelli** [reply]
#74

**Reply by wpbonelli** [reply]
Both the basic and typed parsers now use LALR

**Comment by wpbonelli** [comment]
I think we can implement a [custom xarray backend](https://docs.xarray.dev/en/stable/internals/how-to-add-new-backend.html) to read an input file into a dataset/tree. Internally `open_dataset` can invoke the lark reader.

**Reply by wpbonelli** [reply]
Dedicated issue at #368. This is icing on the cake, not immediately relevant to input file loading

**Comment by wpbonelli** [comment]
We now have a more or less functional input file loading capability with Lark (still some edge cases to work through, but successfully loads most of the test/example models). There are two versions:

- basic: simple grammar, usable for any input file, encodes only the concepts of block, line, etc.
- typed: field-aware grammar per input component, generated from DFNs. still in development.

Related, superseded work

- https://github.com/modflowpy/pyphoenix-project/issues/119
- https://github.com/MODFLOW-ORG/modflow-devtools/pull/193

I did some ASCII file benchmarking. The basic flopy4 loader is roughly competitive with flopy3, with some interesting differences.

- flopy4 is slower to cold start due to import time. one time cost, easy to fix
- flopy4 is marginally faster on most models (127/165) due to lower fixed overhead
- flopy4 is significantly slower on large internal arrays (see token cost, described below)
- flopy4 is significantly faster on large external arrays (see token cost, described below)
- performance is mixed on lists, depending on size/regularity (record vs union/keystring)

The findings show that some of my early reasoning was wrong. The original motivation for the typed loader was multiple: better error (more specific) messages, less manual validation in transformation, and better performance. The first two points hold, but the last is probably wrong. Transformation is only taking a small (<15%) portion of the total load time. Lark lexing/parsing/etc dominates, largely due to the fact we're representing each array value as a distinct token. This blows up the size of the parse tree. It'd be faster and use less memory to parse array and list rows as a single chunk of tokens to load directly with numpy/pandas/etc. The difference between internal and external arrays demonstrates the potential for improvement here: flopy4 already bulk loads external arrays like this, but parses individual tokens for internal arrays, while flopy3 takes a sort of "middle road", parsing lines in both cases. With bulk loading everywhere, I'd expect flopy4 to be faster across the board.

This shows that the basic loader probably suffices for now, and the benefit of the typed loader will mostly be clearer error messages on malformed input files.

Benchmarking infra will be checked in soon
